### DISCUSSION TITLE
textDocument/didClose notification

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1281] [Q&A]
hi should a `textDocument/didClose` notification lead to a `invalidateDocument`or `deleteDocument` in `LangiumDocuments`
i dont see this happening in monaco-languageclient / webworker scenario

**Comment by msujew** [comment]
Usually, in vscode, `textDocument/didClose` doesn't really do anything, since it doesn't change the state of the document - The document is still there, just not open in the editor. If the document changes by closing the editor, `textDocument/didChange` gets called anyway which we correctly apply to the workspace.

In a webworker scenario this semantic might change, as the document could indeed get lost. You should rather listen on `workspace/didDeleteFiles` instead of `textDocument/didClose` in the language server, as that unambiguously denotes that a file has been deleted/removed. Though you might need to emit this notification manually when the document closes. See also https://github.com/eclipse-langium/langium/issues/1230.

**Reply by cdietrich** [reply]
where is this done in langium, where do i need to hook into? or do you think i should send delete file events from the client?

**Reply by msujew** [reply]
Currently, Langium doesn't listen on the `workspace/*` notifications (that's what the issue is tracking). And yes, the event should be sent by the client.

As for hooking into the notification, you can just hook into the connection directly:

```ts
connection.workspace.onDidDeleteFiles(params => ...)
```

**Reply by cdietrich** [reply]
as `"workspace/didChangeWatchedFiles"` is already implemented i workaround using this one.
unfortunately hooking into the events does not seem that easy.

**Reply by msujew** [reply]
Yeah, I forgot that you need to adjust the result of the initialization request by overriding the `LanguageServer#initialize` method... That's why we need a service implementation for this :)

**Reply by cdietrich** [reply]
just for sake of completeness:
doing this in the web-worker (`main-browser.ts` in yo langium)
```
const { shared } = createHelloWorldServices({ connection, ...EmptyFileSystem });

const documentBuilder = shared.workspace.DocumentBuilder;
const documents = shared.workspace.TextDocuments;
const mutex = shared.workspace.MutexLock;

function onDidClose(uri: URI) {
    mutex.lock(token => documentBuilder.update([], [uri], token));
}

documents.onDidClose((event) => {
    onDidClose(URI.parse(event.document.uri));
})
startLanguageServer(shared);
```
seems to do the trick. does it look ok from your pov @msujew

**Reply by msujew** [reply]
Yeah, that's fine. Probably works pretty well from a web IDE perspective, but will break in vscode.

**Reply by cdietrich** [reply]
will it? i assume it would work there too.

**Reply by msujew** [reply]
We don't want to remove a file from the workspace, just because its editor has closed, do we?

**Reply by cdietrich** [reply]
yes but we might want to trigger an update.
maybe this also need to be fixed upstream.
the TextDocuments maintains synced documents and uses that one to remove the dirty open doc.
but also the did open produces a did change.
also on did close of a dirty file langium somehow builds affected files.
did not debug yet why this is working.

**Reply by msujew** [reply]
> also on did close of a dirty file langium somehow builds affected files.
did not debug yet why this is working.

VSCode sends a `textDocument/didChange` when closing a dirty file. We just lean on that behavior.

**Reply by cdietrich** [reply]
and it changes it to empty? or the original content?

**Reply by msujew** [reply]
It changes back to the original content as the dirty state is effectively deleted.
